### ISSUE TITLE
Load/offload optimizer states only when calling optimizer.step()?

### ISSUE THREAD

**Issue by jp1924** [verl-project/verl#5282]
### Feature request

Modify optimizer state load/offload behavior so that optimizer states are loaded into GPU memory only immediately before optimizer.step() and offloaded immediately after. Concretely, add an update_policy that performs:
1) free GPU intermediate tensors to maximize available memory,
2) load optimizer state to GPU,
3) call optimizer.step(),
4) offload optimizer state back to CPU.

Also add a toggle to disable the current ActorRolloutRefWorker-level optimizer load/offload so update_actor does not keep optimizer states resident during the whole update.

### Motivation

Currently ActorRolloutRefWorker.update_actor loads optimizer state up front and keeps it on GPU through the entire actor update (including loss computation padded to max_response_len), which greatly increases peak GPU memory usage. 
Using dynamic batch sizing to avoid large padding is not always possible. 
By loading optimizer state only around optimizer.step and freeing intermediate tensors beforehand, 
peak memory can be reduced, enabling larger models to fit on limited GPUs at the cost of some runtime overhead.

So basically, because data and optimizer states coexist, OOMs happen easily
why not delete data that's no longer needed and reuse that freed memory for the optimizer?

### Your contribution

I can implement the change and submit a PR: add update_policy to perform load-before-step and offload-after-step, add a config flag to disable actor-level persistent optimizer loads, include unit tests and a small benchmark showing peak memory reduction and performance trade-offs, and update docs/examples.

**Comment by Kirrito-k423** [comment]
This is a valuable optimization for memory-constrained scenarios. Your proposal makes sense - loading optimizer states only around `optimizer.step()` can significantly reduce peak memory.

Current memory timeline:
```
loss computation → backward → optimizer.load → step → optimizer.offload
     [data + grads + optimizer_state all on GPU at once]
```

Proposed memory timeline:
```
loss computation → backward → free intermediates → optimizer.load → step → optimizer.offload
     [data only]           [optimizer only]        [data again]
```

Implementation considerations:

1. **Tensor lifecycle**: Need to ensure loss-related tensors are properly freed (del + torch.cuda.empty_cache) before loading optimizer state.

2. **Gradient persistence**: Gradients must be preserved between backward and step. With FSDP, gradients are sharded and may need explicit unsharding.

3. **Performance tradeoff**: Additional load/offload cycles add latency. This is best suited for scenarios where memory, not speed, is the bottleneck.

4. **Config option**: Suggest `actor_rollout_ref.actor.optim.lazy_optimizer_offload=True` to enable this mode.

This would be a great contribution. Consider starting with a prototype on FSDP backend first, then extending to Megatron if needed.

**Comment by jp1924** [comment]
@eric-haibin-lin, @vermouth1992 @wuxibin89
Please, someone come review this!

<img width="184" height="184" alt="Image" src="https://github.com/user-attachments/assets/24702633-1298-4840-816c-fef954376869" />

**Comment by jp1924** [comment]
@eric-haibin-lin, @vermouth1992 @wuxibin89
Please, someone come review this!

<img width="184" height="184" alt="Image" src="https://github.com/user-attachments/assets/24702633-1298-4840-816c-fef954376869" />

**Comment by Kirrito-k423** [comment]
中文：我重新核对了当前 main 分支和相关 PR。这个请求看起来仍然是可行动的功能请求，而不是已经解决的问题：`BaseEngine.train_batch()` 仍然是在 forward/backward 之后调用 `optimizer_step()`，但 `BaseEngineCtx` 的 train-mode 自动 offload 逻辑会在进入训练上下文时按 `optimizer_offload` 加载 optimizer，并在退出上下文时卸载，而不是只在 `optimizer.step()` 周围加载/卸载（见 `verl/workers/engine/base.py:125-127` 和 `verl/workers/engine/base.py:229-264`）。FSDP 实现里也仍然是 `optimizer_step()` 直接调用 `self.optimizer.step()`，optimizer 的 load/offload 在 `to()` 中完成（见 `verl/workers/engine/fsdp/transformer_impl.py:659-698` 和 `verl/workers/engine/fsdp/transformer_impl.py:717-739`）。配置侧目前也只看到 `optimizer_offload` 和 FSDP2 `offload_policy`，没有 `optimizer_lazy_offload` 或类似开关（见 `verl/trainer/config/engine/fsdp.yaml:10-19`）。

同时，#5657 曾经实现过这个方向并引用了本 issue，但该 PR 已关闭且未合并，review 状态仍是 `REVIEW_REQUIRED`，所以我没有找到可以支持关闭本 issue 的证据。比较合适的下一步是基于当前 new worker/FSDP engine 重新整理或重开一个实现 PR，并补上梯度保留、释放中间张量、峰值显存/性能对比和测试覆盖。

English: I checked current main and the related PR again. This still looks like an actionable feature request rather than something already fixed. `BaseEngine.train_batch()` still runs forward/backward before `optimizer_step()`, while `BaseEngineCtx` loads/offloads optimizer state at train-mode context entry/exit when `optimizer_offload` is enabled, instead of wrapping only `optimizer.step()` (`verl/workers/engine/base.py:125-127`, `verl/workers/engine/base.py:229-264`). In the FSDP engine, `optimizer_step()` still directly calls `self.optimizer.step()`, and optimizer load/offload is handled by `to()` (`verl/workers/engine/fsdp/transformer_impl.py:659-698`, `verl/workers/engine/fsdp/transformer_impl.py:717-739`). The current FSDP config exposes `optimizer_offload` and FSDP2 `offload_policy`, but I did not find an `optimizer_lazy_offload`/similar toggle (`verl/trainer/config/engine/fsdp.yaml:10-19`).

Also, #5657 implemented this idea and referenced this issue, but it is closed without being merged and still shows `REVIEW_REQUIRED`, so I do not see enough evidence to close this issue as resolved. A good next step would be to revive/rebase the implementation against the current new worker/FSDP engine, with coverage for gradient preservation, intermediate tensor cleanup, peak-memory/performance trade-off, and tests.

This response was generated by a scheduled Codex GPT5.5 issue processing task.
